### PR TITLE
Add codemods for Version 3.0

### PR DESCRIPTION
**Pull request by alex-ju**
### :pushpin: Summary

[As agreed during the engineering sync](https://docs.google.com/document/d/1twnk39hZ7p8oZYrld_FYsdiuUb6yn_weA46KnhF4nu4/edit#bookmark=id.1wyi4r3swd72), I've started to prepare the codemods for the breaking changes we're planning to release in [version 3.0](https://github.com/hashicorp/design-system/pull/1634).

### :hammer_and_wrench: Detailed description

- I added one codemod for the `Dropdown` changes. If we agree this is the right format, I will add the remaining two (for the `MaskedInput` and the `Form::RadioCard`).
- It seems to be the norm to publish these codemods as a separate package, so I placed them accordingly.

As always open for feedback on both the format and the location.

***

### 👀 Component checklist

- [x] Percy was checked for any visual regression
- [x] A11y tests have been run locally
- [x] A changelog entry was added via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

### CODE REVIEW

File: packages/codemods/package.json
Comment (didoo) [inline packages/codemods/package.json:9]: 🤔

File: packages/codemods/transforms/v3/dropdown/README.md
Comment (didoo) [inline packages/codemods/transforms/v3/dropdown/README.md:1]: we should probably be more verbose and call the folder with the full semver value `3.0.0` (in case we have a 3.1.0 codemod, we never know)

File: packages/codemods/transforms/v3/dropdown/README.md
Comment (didoo) [inline packages/codemods/transforms/v3/dropdown/README.md:1]:
```suggestion
# v3.0.0/dropdown
```

File: packages/codemods/transforms/v3/dropdown/README.md
Comment (didoo) [inline packages/codemods/transforms/v3/dropdown/README.md:18]: I would do something like this to make it clearer what options they have (I can't write the exact markdown, because of the multiple backticks, but I hope it gives you the gist of it):

## Usage

### Using `npx`
```
npx @hashicorp/design-system-codemods v3.0.0/dropdown path/to/single/file.hbs
```
or
```
npx @hashicorp/design-system-codemods v3.0.0/dropdown path/to/some/glob/**/*.hbs
```

### Global installation
```
yarn global add @hashicorp/design-system-codemods
```
and then
```
@hashicorp/design-system-codemods v3.0.0/dropdown path/to/single/file.hbs
@hashicorp/design-system-codemods v3.0.0/dropdown path/to/some/glob/**/*.hbs
```

### Local installation
```
node ./bin/cli.js v3.0.0/dropdown path/to/single/file.hbs
node ./bin/cli.js v3.0.0/dropdown path/to/some/glob/**/*.hbs
```

Comment (didoo) [reply packages/codemods/transforms/v3/dropdown/README.md:18]: (we may also want to limit the proposed approaches to only one or two)

File: packages/codemods/transforms/v3/dropdown/__testfixtures__/basic.input.hbs
Comment (didoo) [inline packages/codemods/transforms/v3/dropdown/__testfixtures__/basic.input.hbs:1]: I don't know how these test work, but should we also have a version where the dropdown is already using the `@listPosition="bottom-right"` (for example) and the test makes sure we don't touch it, the output remains the same?

File: packages/codemods/transforms/v3/dropdown/__testfixtures__/basic.input.hbs
Comment (didoo) [inline packages/codemods/transforms/v3/dropdown/__testfixtures__/basic.input.hbs:1]: I would probably use a much simpler code snippet, with a single child (unless we're worried it may modify the children too, in which case I would add one per kind of item)

File: packages/codemods/transforms/v3/dropdown/README.md
Comment (didoo) [reply packages/codemods/transforms/v3/dropdown/README.md:18]: Or maybe here we mention only one of the methods (with the two file paths) and we link to the `packages/codemods/README.md` for details about other possible methods.

File: packages/codemods/README.md
Comment (didoo) [inline packages/codemods/README.md:5]: See comment below (sorry, I saw this after the comment; I don't approach the code review in linear way, at times)

File: packages/codemods/README.md
Comment (didoo) [inline packages/codemods/README.md:27]: We should probably group them together (unless this is automated, which I suspect)
```suggestion
<!--TRANSFORMS_START-->
### v3.0.0
* [v3.0.0/dropdown](transforms/v3.0.0/dropdown/README.md)
<!--TRANSFORMS_END-->
```

File: packages/codemods/.github/workflows/ci.yml
Comment (didoo) [inline packages/codemods/.github/workflows/ci.yml:36]: What are we testing here? 🤔

File: packages/codemods/transforms/.gitkeep
Comment (didoo) [inline packages/codemods/transforms/.gitkeep:1]: maybe this can be deleted?

**Review by didoo — COMMENTED** [review]
This approach works for me (I have left only minor comments) so feel free to continue along this line. 

One general consideration: should we move some of the codemods from Cloud UI to here, so they can be used as reference/example? (only the ones that are generic, not specific to Cloud UI).

File: packages/codemods/transforms/v3/dropdown/index.js
Comment (didoo) [inline packages/codemods/transforms/v3/dropdown/index.js:18]: I find this pattern ([used in Cloud UI](https://github.com/hashicorp/cloud-ui/blob/main/codemods/transforms/hds-migration/core/external-link/index.js#L84-L85)) a bit clearer/explicit:
```suggestion
            const updatedListPositionAttr = b.attr('@listPosition', b.text(`bottom-${attr.value.chars}`));
            outputAttrs.push(updatedListPositionAttr);
```

File: packages/codemods/transforms/v3/dropdown/__testfixtures__/basic.input.hbs
Comment (alex-ju) [reply packages/codemods/transforms/v3/dropdown/__testfixtures__/basic.input.hbs:1]: Good call, yes, I can add a test to make sure we don't accidentally touch code we don't want to touch.

File: packages/codemods/transforms/v3/dropdown/__testfixtures__/basic.input.hbs
Comment (alex-ju) [reply packages/codemods/transforms/v3/dropdown/__testfixtures__/basic.input.hbs:1]: yes, that works for me, just wanted to make sure we have node elements in the block content

File: packages/codemods/transforms/v3/dropdown/README.md
Comment (alex-ju) [reply packages/codemods/transforms/v3/dropdown/README.md:1]: I worked under the assumption that a code mode is required only for API changes. If we made any change that would require a change in our consumer's codebase it means it's a breaking change. This is why I only see us adding code mods for major releases.

The reason I named it generically v3 (or it can be 3.x.x) is that these code mods can be used to move from 2.7 to 3.2 or even 2.7 to 4.2 (by consequently running the v3 mods and then the v4 mods). Happy to open this to a wider discussion.

File: packages/codemods/README.md
Comment (alex-ju) [reply packages/codemods/README.md:27]: yes, it's unfortunately automated by running the `update-docs` command you touched on above.

File: packages/codemods/.github/workflows/ci.yml
Comment (alex-ju) [reply packages/codemods/.github/workflows/ci.yml:36]: This was the more or less auto-generated CI config. I would lean towards disabling these jobs and only running the tests locally when writing the code mods. I think the overhead is quite big and don't think it's worth running them in CI.

File: packages/codemods/transforms/.gitkeep
Comment (alex-ju) [reply packages/codemods/transforms/.gitkeep:1]: yes, indeed

**Comment by alex-ju** [comment]
> One general consideration: should we move some of the codemods from Cloud UI to here, so they can be used as reference/example? (only the ones that are generic, not specific to Cloud UI).

it's probably worth porting those here when we find the time considering we have 2 products on v1.x; there are a few others (I'm aware of a few written by me for v2 where I was asked not to merge them in `cloud-ui` so I dropped the commits before bumping HDS to v2) that may be worth reviving as well.

File: packages/codemods/transforms/v3/dropdown/README.md
Comment (alex-ju) [reply packages/codemods/transforms/v3/dropdown/README.md]: Updated with a simplified version, suggesting npx for consumers and local binary for maintainers. Agree there were to many options

File: packages/codemods/transforms/v3/radio-card/index.js
Comment (didoo) [inline packages/codemods/transforms/v3/radio-card/index.js:34]: does this apply to any other kind of node?

File: packages/codemods/transforms/v3/radio-card/index.js
Comment (didoo) [inline packages/codemods/transforms/v3/radio-card/index.js:3]: according to the changelog https://github.com/hashicorp/design-system/pull/1634/files#diff-fa1aa341f6153e2733520f2d8227380e02de23c96082cac195a30f4ddde9fbe7 also a `@maxWidth` should be applied. If this is the case, how is the codemod going to take care of it?

**Review by didoo — COMMENTED** [review]
Added a few comments (apart from that, probably ready to be approved unless you have to add more codemods; on my side, I am going to push a codemod for the `SideNav` right now on this branch)

File: packages/codemods/transforms/v3/radio-card/index.js
Comment (alex-ju) [reply packages/codemods/transforms/v3/radio-card/index.js]: 🤦 I forgot the node name – amended now

File: packages/codemods/transforms/v3/radio-card/index.js
Comment (didoo) [reply packages/codemods/transforms/v3/radio-card/index.js:3]: Maybe something to add to the README for this codemod?

Comment (alex-ju) [reply packages/codemods/transforms/v3/radio-card/index.js:3]: That's a good shout – added an extra note to the `v3/radio-card` instructions

**Review by didoo — APPROVED** [review]
👍

**Pull request merged**
